content/purgatorio/launch-and-mantain-the-system/monitoring-usage-and-behavior.md

```diff
@@ -1,5 +1,142 @@
+---
+
+# Index 
+
+- [Trusting is Good Not Trusting is Better](#Trusting-is-Good-Not-Trusting-is-Better)
+- [Monitoring the Behavior](#Monitoring-the-Behavior)
+- [Monitoring the Usage](#Monitoring-the-Usage)
+- [Conclusions](#Conclusions)
+
+
+## Trusting is Good Not Trusting is Better
+
+Application monitoring is a _key_ part of running software in production. 
+
+Without it, the only way of finding an issue is through sheer luck or because a client has reported it. 
+
+Both of which are less than ideal, to say the least!
+
+**You wouldn't deploy an application without monitoring, so why do it for Machine Learning models?**
+
+So let's Start!
+
+The first resource you should go through is the following: 
+
+- [How to Monitor Machine Learning Models](https://christophergs.com/machine%20learning/2020/03/14/how-to-monitor-machine-learning-models/)
+
+It's a very detailed and comprehensive blog post and it addresses these topics:
+
+- What makes ML systems monitoring hard?
+- How can we monitor the usage and behavior of the model? 
+- What are the key metrics to track and alert on?
+- What are the key principles for monitoring the ML system?
+
+Once you are done with the previous blog post, you can refer to the related chapter of the book [Machine Learning Engineering](http://www.mlebook.com/wiki/doku.php) for a more detailed guide (strongly recommended buy, but you can read it for free online).
+
+- [Chapter 8 - Model Serving and Monitoring](https://www.dropbox.com/s/nj23m5myng9gram/Chapter8.pdf?dl=0)
+
+See also:
+ - [There are two very different ways to deploy ML models, here’s both](https://towardsdatascience.com/there-are-two-very-different-ways-to-deploy-ml-models-heres-both-ce2e97c7b9b1)
+
+With these two resources, you should understand broadly the reasons and challenges of monitoring the Machine Learning models in production, and you should have reasonable strategies to tackle them. 
+
+As you know, the two main challenges regarding the model monitoring are:
+
+- How does the system behave?
+- How is the system used?
+
+Let's see some specific resources for these two challenges.
+
+## Monitoring the Behavior
+
+When monitoring the behavior of the ML model in production, you should consider many aspects:
+
+
+- **Data in:**
+
+    - Monitor whether the data you're processing looks like the data you trained on. ([data drift challenge](https://www.cmswire.com/digital-experience/data-drift-what-it-is-and-how-to-avoid-it/)) E.g., use simple (comparatively interpretable) distributional models to try to track whether data looks "sufficiently" similar.
+    
+    See:
+    - [Calculate Similarity — the most relevant Metrics in a Nutshell](https://towardsdatascience.com/calculate-similarity-the-most-relevant-metrics-in-a-nutshell-9a43564f533e)
+    - [How to measure similarity or dissimilarity between two data set?](https://www.researchgate.net/post/How_to_measure_similarity_or_dissimilarity_between_two_data_set)
+
```

Comment:
Usually, a basic descriptive stats with boundary checks can be a good start, there is a good library to just check that - https://github.com/EricSchles/drifter_ml (highly recommend to add this somewhere)

Reply:
Specific: we can also find out if the distribution of the live-data or evaluation data matches that of the training set (or even the held-out test/evaluation set), one specific method that comes to mind is `Kolmogorov-Smirnov` test (highly recommend to include a link to this somewhere. Its a big specific but quite relevant.